### ISSUE TITLE
diagnostic(p3): inspect rollout.rewards signal — is the per-step reward actually varying?

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#190]
## Hypothesis (H1)

**The reward signal entering PPO is degenerate or near-constant**, regardless of agent actions. If so, no PPO hyperparameter tuning matters because there is no learnable gradient.

## Why this is the cheapest informative test

After three iterations of PPO ablation (#145, #174, #183), the trained-policy reward sits at iter-0's value (~290) over 500 iters with stable gradients, prevented entropy collapse, and value-loss dominance crushed. PPO's optimizer is healthy but reward doesn't move. The simplest explanation is that PPO has no signal to learn from.

## What to measure

In a single PPO cell on \`default\` at λ=0 (use L1_norm config, seed 42 — matches one already in \`runs/p3_183_phase3/L1_norm/default/lambda_0e0/seed_42/\` if you don't want to retrain):

For the most recent rollout, dump:

1. \`rollout.rewards.shape\` — confirm it's \`(num_agents, T)\` or whatever the layout actually is
2. \`rollout.rewards.mean(axis=-1)\` per agent — mean per-step reward
3. \`rollout.rewards.std(axis=-1)\` per agent — std per-step reward
4. \`rollout.rewards.std() / rollout.rewards.mean()\` — coefficient of variation
5. Histogram of per-step reward values across the rollout (10–20 bins)
6. Correlation between per-step reward and per-step action choice (do different actions produce noticeably different rewards?)

## Acceptance criteria

- [ ] CV (σ/μ) is reported per-agent. If CV < 0.05, that's the smoking gun: rewards are near-constant.
- [ ] Per-step reward distribution shown (mean, std, p10, p50, p90).
- [ ] Action-reward correlation reported: do different actions correlate with materially different rewards? If r² < 0.01, actions don't matter.
- [ ] If signal IS varying, the histogram shows the spread — useful even on a negative result for H1 (rules it out, points at H2 or H4).

## Implementation notes

- The rollout objects live in \`bucket_brigade/training/joint_trainer.py::collect_rollout\` (or wherever \`rollout = trainer.collect_rollout(...)\` originates in \`experiments/p3_specialization/train.py:278\`).
- The Python env's per-step rewards include team-reward + per-agent ownership deltas (see closed issue #135 — Python folds in team_reward_house_survives * saved_fraction every step, unlike Rust core which only does it at episode end). So per-step reward is dominated by team-level signal; per-agent variance comes from ownership deltas.
- A 10–20 line standalone script in \`experiments/p3_specialization/diagnostics/\` is the right deliverable — no train.py changes, no PR to merge.

## Out of scope

- Fixing the reward signal if it IS degenerate — that's a downstream issue (likely H4)
- Retraining anything — use existing checkpoints if possible
- Testing other scenarios (chain_reaction) — start with default, expand if H1 confirms

## Why not just look at the metrics.json

\`metrics.json\` logs the **per-iteration mean team reward** (a scalar), not the per-step distribution within a rollout. H1 is asking about the distribution PPO actually sees, which is lost in the aggregation.

## References

- #183 (Phase 3 negative result that motivated the rethink)
- #135 (closed, but documents the Python env's per-step reward decomposition)
- The conversation that led to this: PPO-tuning ladder is exhausted; next question is whether the signal entering PPO is informative

## Estimated effort

~30 min. Cheapest of the four diagnostic issues.

**Comment by rjwalters** [comment]
## Curator Enhancement (2026-05-15)

### Verified Technical Claims

**1. Rollout structure — important correction.** `rollout.rewards` is **not** a single ndarray with `shape=(num_agents, T)`. From `bucket_brigade/training/joint_trainer.py:71-91` (the `RolloutBuffer` dataclass) and the construction at `joint_trainer.py:223-273`:

- `rollout.rewards` is `Dict[int, torch.Tensor]`, one entry per agent.
- Each `rollout.rewards[i]` is a 1-D `torch.Tensor` of shape `[T]` (where `T == rollout_steps`).
- Internally collected as `rewards = np.zeros((self.num_agents, num_steps), dtype=np.float32)` at line 231, with `rewards[i, t] = rew[i]` at line 246, then split into the dict at line 268.

The diagnostic script will need to **stack the dict back into an ndarray** before computing distributional stats, e.g. `R = torch.stack([rollout.rewards[i] for i in range(N)]).cpu().numpy()  # shape (N, T)`.

**2. Rollout collection site in train.py:** confirmed at **`experiments/p3_specialization/train.py:478`** (`rollout = trainer.collect_rollout(cfg.rollout_steps)`); the issue's `:278` reference is stale.

**3. Existing cell location — path correction.** The issue cites `runs/p3_183_phase3/L1_norm/default/lambda_0e0/seed_42/`. That path does **not exist**. The actual layout is `experiments/p3_specialization/runs/{scenario}/{lambda_NeN}/seed_{N}/`. The matching cell is:

```
experiments/p3_specialization/runs/default/lambda_0e0/seed_42/
├── config.json          # seed=42, scenario=default, lambda_red=0.0, num_iterations=50, rollout_steps=2048, num_agents=4, hidden_size=64, lr=3e-4, ppo_epochs=4, minibatch_size=256, entropy_coef=0.01, action_dims=[10, 2]
├── metrics.json         # per-iter scalars; mean_step_reward_team is iter-0 ≈ 295.69 (matches the "stuck near 290" observation)
├── policies/
│   ├── agent_0.pt
│   ├── agent_1.pt
│   ├── agent_2.pt
│   └── agent_3.pt
└── dropout_results.json
```

Note: this cell ran **50 iterations** (`num_iterations: 50`), not 500. The 500-iter Phase 3 results referenced in #183 are not in this tree under that filename — the diagnostic can still use seed_42 since H1 is about the per-step reward signal from a converged-or-stuck policy, and the 50-iter checkpoint is already plateaued. If a 500-iter checkpoint exists elsewhere on sandbox-1, prefer that; otherwise seed_42's 50-iter policies are a valid stand-in.

**4. Python env reward decomposition** (per-step, confirmed at `bucket_brigade/envs/bucket_brigade_env.py:248-298`): each step's per-agent reward = `team_reward` (shared across agents) + work/rest term (±cost_to_work_one_night or +0.5) + ownership-delta term (event-triggered: ±reward_own_house_survives / penalty_own_house_burns). For `default`: `team_reward_house_survives=100.0`, `team_penalty_house_burns=100.0`, `cost_to_work_one_night=0.5`, `reward_own_house_survives=1.0`, `penalty_own_house_burns=2.0`. This confirms the issue's note from #135: per-step reward is dominated by the shared team term.

### CV < 0.05 threshold — needs reconsideration

The issue proposes `CV < 0.05 = smoking gun`. Given the env's structure, this threshold is **probably too aggressive** and may flag a non-degenerate signal as degenerate:

- The team component is `100 * saved_fraction`. Once most houses are saved (which is the plateau regime, mean_team_reward ≈ 290 across a 2048-step rollout where one episode lasts ~12+ nights with min_nights=12), `saved_fraction` may sit near a constant value most of the time, with brief transitions. Mean per-step reward of ~70-75 per agent (290/4) is plausible.
- Std of per-step reward is driven by: (a) team component changes when houses transition state (rare relative to T=2048), (b) work/rest swings of order 0.5-1.0 every step, (c) ownership deltas (event-triggered, sparse).
- A CV of std/mean = 0.05 against mean ≈ 70 gives std ≈ 3.5 — easily achievable just from the work/rest term toggling every step, even with a fully degenerate policy.

**Recommended methodology refinements:**

1. **Decompose the reward** rather than just computing aggregate CV. The env code at lines 248-298 cleanly separates team / work-rest / ownership components — log each independently. The most informative quantity is the **std of the action-controllable portion** (work/rest + ownership), not the total.
2. **Add a per-component CV** (CV_team, CV_work, CV_ownership) in addition to the aggregate. If `CV_work` or `CV_ownership` are zero or near-zero, that's far more diagnostic than aggregate CV — it means the policy isn't taking action-distinguishable steps.
3. **Action-reward correlation** (item 6 in the original AC) is the strongest test: regress per-step reward on per-step action and report R². If R² < 0.01 holds, *that* is the smoking gun. Keep this AC.
4. **Compare to a fresh random-init agent** on the same rollout setup — if random-init produces the same flat distribution, that confirms the env, not the policy, is degenerate. (Cheap: ~10 extra lines of script.)

### Concrete Script Outline (~40 lines, standalone)

Target location: `experiments/p3_specialization/diagnostics/inspect_rollout_rewards.py`

```python
"""H1 diagnostic: inspect per-step rollout.rewards distribution.

Loads trained policies from a P3 cell, runs one rollout, dumps per-step
reward statistics to confirm/rule out reward-signal degeneracy.
"""
import json
from pathlib import Path
import numpy as np
import torch

from bucket_brigade.envs.bucket_brigade_env import BucketBrigadeEnv
from bucket_brigade.envs.scenarios_generated import get_scenario_by_name
from bucket_brigade.training.joint_trainer import JointPPOTrainer, flatten_dict_obs

CELL = Path("experiments/p3_specialization/runs/default/lambda_0e0/seed_42")
cfg = json.loads((CELL / "config.json").read_text())

scenario = get_scenario_by_name(cfg["scenario"], num_agents=cfg["num_agents"])
env_fn = lambda: BucketBrigadeEnv(scenario=scenario)
probe = env_fn(); obs_dim = flatten_dict_obs(probe.reset(seed=cfg["seed"])).shape[0]

trainer = JointPPOTrainer(
    env_fn=env_fn, num_agents=cfg["num_agents"], obs_dim=obs_dim,
    action_dims=cfg["action_dims"], hidden_size=cfg["hidden_size"],
    lr=cfg["lr"], ppo_epochs=cfg["ppo_epochs"],
    minibatch_size=cfg["minibatch_size"], entropy_coef=cfg["entropy_coef"],
    device=cfg["device"], seed=cfg["seed"],
)
for i in range(cfg["num_agents"]):
    trainer.policies[i].load_state_dict(torch.load(CELL / f"policies/agent_{i}.pt"))

rollout = trainer.collect_rollout(cfg["rollout_steps"])

# Stack dict[int, Tensor[T]] -> ndarray[N, T]
R = torch.stack([rollout.rewards[i] for i in range(cfg["num_agents"])]).cpu().numpy()
A = torch.stack([rollout.actions[i] for i in range(cfg["num_agents"])]).cpu().numpy()  # [N, T, 2]

print(f"R.shape = {R.shape}")
for i in range(cfg["num_agents"]):
    r = R[i]; cv = r.std() / (abs(r.mean()) + 1e-9)
    p10, p50, p90 = np.percentile(r, [10, 50, 90])
    print(f"agent_{i}: mean={r.mean():.3f} std={r.std():.3f} CV={cv:.4f} "
          f"p10/50/90={p10:.2f}/{p50:.2f}/{p90:.2f}")
    # Action-reward correlation (work/rest dim only — discrete bivariate)
    work = A[i, :, 1]  # 0=rest, 1=work
    r_rest, r_work = r[work == 0], r[work == 1]
    if len(r_rest) and len(r_work):
        # R^2 of a simple binary regressor (mean reward conditional on action)
        ss_total = ((r - r.mean()) ** 2).sum()
        pred = np.where(work == 0, r_rest.mean(), r_work.mean())
        ss_res = ((r - pred) ** 2).sum()
        r2 = 1 - ss_res / (ss_total + 1e-9)
        print(f"  r_rest={r_rest.mean():.3f} (n={len(r_rest)}) "
              f"r_work={r_work.mean():.3f} (n={len(r_work)}) R²={r2:.4f}")

# Histogram (text-mode is fine; matplotlib optional)
hist, edges = np.histogram(R.flatten(), bins=15)
print("\nHistogram of all per-step rewards:")
for h, lo, hi in zip(hist, edges[:-1], edges[1:]):
    bar = "#" * int(60 * h / hist.max())
    print(f"  [{lo:7.2f}, {hi:7.2f}) {h:6d} {bar}")
```

Optional extension (recommended): repeat with a fresh `JointPPOTrainer` (random init, no `load_state_dict`) and print the same stats — gives a baseline to compare the trained-policy CV against.

### Affected Files

- `experiments/p3_specialization/diagnostics/inspect_rollout_rewards.py` — **NEW** standalone script (the deliverable)
- No source-file modifications. No `train.py` changes. No PR to merge.

### Test Plan

- [ ] Manual verification: run `uv run python experiments/p3_specialization/diagnostics/inspect_rollout_rewards.py` from repo root; script completes without exceptions in <2 min on CPU.
- [ ] Output reports per-agent: mean, std, aggregate CV, p10/p50/p90 of per-step reward.
- [ ] Output reports per-agent R² of reward-vs-(work/rest)-action.
- [ ] Output includes histogram (text or saved PNG) of the per-step reward distribution.
- [ ] **Interpretation step** (documented in the script's printout, not just the issue): the user manually inspects the output and makes a call. The interpretive rubric:
  - If aggregate CV < 0.05 **and** action-reward R² < 0.01 → H1 confirmed (degenerate signal).
  - If aggregate CV is low but action-reward R² is meaningful (>0.05) → H1 is *not* the issue; the signal varies with action but PPO isn't capturing it. Points at H2/H3/H4.
  - If CV is high and R² is meaningful → H1 ruled out, reward signal is informative. Move to other hypotheses.

### Acceptance Criteria

Carried forward from the original issue with refinements:

- [ ] Aggregate CV (σ/μ) reported per-agent.
- [ ] **Per-component CV** also reported, or at minimum the issue commenter documents why aggregate CV alone is sufficient (since team term dominates per-step variance, aggregate CV alone may be uninformative — see methodology note above).
- [ ] Per-step reward distribution shown: mean, std, p10, p50, p90, plus histogram.
- [ ] Action-reward R² reported (binary work/rest split is sufficient; the 10-way location action is also welcome but not required).
- [ ] If signal is varying, the histogram is included in the diagnostic output even on a negative result for H1 (rules H1 out, points downstream).
- [ ] **Standalone deliverable**: lives in `experiments/p3_specialization/diagnostics/`, runs as `uv run python ...`, no train.py / source modifications.
- [ ] No PR. Diagnostic output (logs, plots if any) committed to the diagnostics dir or pasted into a follow-up comment on this issue.

### Scope Guards

- ❌ No modifications to `train.py`, `joint_trainer.py`, `bucket_brigade_env.py`, or any other source file.
- ❌ No PR. This is a diagnostic-output deliverable, not a code-change one.
- ❌ No retraining. Reuse the existing seed_42 checkpoint.
- ❌ No testing of other scenarios (chain_reaction, trivial_cooperation). `default` only — expand only if H1 confirms on default.
- ❌ No fixing the reward signal if degenerate — out of scope (downstream, see #193).

### References

- Sibling diagnostics: #191 (per-agent attribution audit), #192 (random baseline re-derivation), #193 (reward-shaping proposal)
- Phase-3 negative result that motivated this: #183
- Per-step reward folding (Python env) precedent: #135 (closed)
- Phase 2 ablation that exhausted the PPO-tuning ladder: #174 (closed)
